network/netplan/netplan.go

```diff
@@ -0,0 +1,245 @@
+	Via    string `yaml:"via,omitempty"`
+	Metric int    `yaml:"metric,omitempty"`
+}
+
+type Network struct {
+	Version   int                 `yaml:"version"`
+	Renderer  string              `yaml:"renderer,omitempty"`
+	Ethernets map[string]Ethernet `yaml:"ethernets,omitempty"`
+	Wifis     map[string]Wifi     `yaml:"wifis,omitempty"`
+	Bridges   map[string]Bridge   `yaml:"bridges,omitempty"`
+	Routes    []Route             `yaml:"routes,omitempty"`
+}
+
+type Netplan struct {
+	Network         Network `yaml:"network"`
+	sourceDirectory string
+	sourceFiles     []string
+	backedFiles     map[string]string
+	writtenFile     string
+}
+
+// BridgeDeviceById takes a deviceId and creates a bridge with this device
+// using this devices config
+func (np *Netplan) BridgeEthernetById(deviceId string, bridgeName string) (err error) {
+	ethernet, ok := np.Network.Ethernets[deviceId]
+	if !ok {
+		return errors.NotFoundf("Device with id %q", deviceId)
+	}
+	for bName, bridge := range np.Network.Bridges {
+		for _, i := range bridge.Interfaces {
+			if i == deviceId {
+				// The device is already properly bridged, we're not doing anything
+				if bridgeName == bName {
+					return nil
+				} else {
+					return errors.AlreadyExistsf("Device %q is already bridged in bridge %q instead of %q", deviceId, bName, bridgeName)
+				}
+			}
+		}
+		if bridgeName == bName {
+			return errors.AlreadyExistsf("Bridge named %q", bridgeName)
+		}
+	}
+	// copy aside and clear the IP settings from the original Ethernet device, except for MTU
+	intf := ethernet.Interface
+	ethernet.Interface = Interface{MTU: intf.MTU}
+	// create a bridge
+	if np.Network.Bridges == nil {
+		np.Network.Bridges = make(map[string]Bridge)
+	}
+	np.Network.Bridges[bridgeName] = Bridge{
+		Interfaces: []string{deviceId},
+		Interface:  intf,
+	}
+	np.Network.Ethernets[deviceId] = ethernet
+	return nil
+}
+
+func Unmarshal(in []byte, out interface{}) (err error) {
+	return goyaml.UnmarshalStrict(in, out)
+}
+
+func Marshal(in interface{}) (out []byte, err error) {
+	return goyaml.Marshal(in)
+}
+
+// readYamlFile reads netplan yaml into existing netplan structure
+// TODO(wpk) 2017-06-14 When reading files sequentially netplan replaces single
+// keys with new values, we have to simulate this behaviour.
+func (np *Netplan) readYamlFile(path string) (err error) {
+	contents, err := ioutil.ReadFile(path)
+	if err != nil {
+		return err
+	}
+	err = Unmarshal(contents, np)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// ReadDirectory Reads the contents of a netplan directory and
+// returns complete config.
+func ReadDirectory(dirPath string) (np Netplan, err error) {
+	fileInfos, err := ioutil.ReadDir(dirPath)
+	if err != nil {
+		return np, err
+	}
+	np.sourceDirectory = dirPath
+	for _, fileInfo := range fileInfos {
+		if !fileInfo.IsDir() && strings.HasSuffix(fileInfo.Name(), ".yaml") {
+			np.sourceFiles = append(np.sourceFiles, fileInfo.Name())
+		}
+	}
```

Comment:
We should have concrete tests that we read these in the exact order that Netplan will read them.